### PR TITLE
Version 0

### PR DESCRIPTION
**Pull request by dlanger**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+a language-agnostic A/B testing framework.
+
+This is not a full-fledged client (it relies on ``sixpack-py`` for the actual connection); it's a wrapper
+to make using ``sixpack-py` more friendly and declarative.
+
+
+Install
+-------
+
+.. code::
```
Comment (ashchristopher) [inline README.rst:15]: Markdown or Rst? Why choose rst?

File: djsixpack/djsixpack.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+from sixpack import sixpack
+from django.conf import settings
+
+RE_FIRST_CAP = re.compile('(.)([A-Z][a-z]+)')
+RE_ALL_CAP = re.compile('([a-z0-9])([A-Z])')
+RE_TEST_NAME = re.compile('_test$')
+
+
+class AlternativesAttributeSetterMeta(type):
```
Comment (ashchristopher) [inline djsixpack/djsixpack.py:11]: Why use a metaclass?

File: runtests.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import os
+import sys
+
+from django.conf import settings
+
+
+if not settings.configured:
+    settings_dict = dict(
+        INSTALLED_APPS=(
+			"djsixpack",
```
Comment (ashchristopher) [inline runtests.py]: Simple formatting issue.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    include_package_data=True,
+    zip_safe=True,
+    packages=find_packages(),
+    url='https://github.com/waveaccounting/django-sixpack',
+    license='LICENSE',
+    description='A django-friendly wrapper for sixpack-py',
+    long_description=open('README.rst').read(),
+    install_requires=["Django>=1.4", "sixpack-client==1.0.0"],
+    tests_require=["Django>=1.4", "mock==1.0.1", "sixpack-client==1.0.0"],
+    test_suite='runtests.runtests',
```
Comment (ashchristopher) [inline setup.py]: Just out of curiosity, can you run `python setup.py test`?

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from setuptools import setup, find_packages
+
+setup(
+    name='django-sixpack',
+    version=__import__('djsixpack').__version__,
+    author=open('AUTHORS').read(),
+    author_email='opensource@waveaccounting.com',
```
Comment (ashchristopher) [inline setup.py]: If you are the primary contributor, I would put your name here. The AUTHORS file is available for additional contributors. I think it is important you (and anyone else who does something for OS) get the proper recognition.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    include_package_data=True,
+    zip_safe=True,
+    packages=find_packages(),
+    url='https://github.com/waveaccounting/django-sixpack',
+    license='LICENSE',
+    description='A django-friendly wrapper for sixpack-py',
+    long_description=open('README.rst').read(),
+    install_requires=["Django>=1.4", "sixpack-client==1.0.0"],
+    tests_require=["Django>=1.4", "mock==1.0.1", "sixpack-client==1.0.0"],
```
Comment (mwarkentin) [inline setup.py]: This could be `tests_require=['mock==1.0.1"] + install_requires`

File: runtests.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import os
```
Comment (mwarkentin) [inline runtests.py:1]: Would it be worth switching from naming this file `runtests.py` to `unittest.py`? It sounds like CircleCI will automatically run that if it's in the repo: https://circleci.com/docs/language-python#testing

File: README.rst (hunk 1 above)
Comment (dlanger) [inline README.rst:15]: PyPI expects the `README` to be in RST (I've seen instructions with `setup.cfg` to make it work with Markdown, but have never never successfully done it)

File: setup.py (hunk 4 above)
Comment (dlanger) [inline setup.py]: Haven't tried yet - once I get content down, then I'm going to battle packaging some more.

File: setup.py (hunk 6 above)
Comment (dlanger) [inline setup.py]: I don't think that exact syntax works, but I've refactored to hit the spirit of what you're going for.

File: setup.py (hunk 5 above)
Comment (dlanger) [inline setup.py]: Cool - I wasn't sure if we had a convention for this.

File: runtests.py (hunk 7 above)
Comment (dlanger) [inline runtests.py:1]: Once it's public, I'm going to try and get travis-ci running on it (I haven't looked yet how to get that picked up right).

File: runtests.py (hunk 3 above)
Comment (dlanger) [inline runtests.py]: Done

File: djsixpack/djsixpack.py (hunk 2 above)
Comment (dlanger) [inline djsixpack/djsixpack.py:11]: So that we can expose the alternatives as class-level attributes automatically without instantiating an experiment.

Comment (ashchristopher) [inline djsixpack/djsixpack.py:11]: Maybe you can show me in person - not sure I follow.

File: README.rst (hunk 1 above)
Comment (nickpresta) [inline README.rst:15]: I also have yet to find a way to get PyPI to work with displaying the readme as inline markdown.

File: MANIFEST.in [hunk 8]
```diff
@@ -0,0 +1,3 @@
+include README.rst
+include AUTHORS
+include LICENSE
```
Comment (nickpresta) [inline MANIFEST.in:3]: This file doesn't exist.

File: djsixpack/__init__.py [hunk 9]
```diff
@@ -0,0 +1 @@
+__version__ = "0.1.0"
```
Comment (nickpresta) [inline djsixpack/__init__.py]: Should be `1.0.0` if this is indeed the first stable release.

File: djsixpack/djsixpack.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        s1 = RE_FIRST_CAP.sub(r'\1_\2', self.__class__.__name__)
+        name = RE_ALL_CAP.sub(r'\1_\2', s1).lower()
+        return RE_TEST_NAME.sub('', name)
+
+    def get_client_id(self, instance):
+        if not self.unique_attr:
+            raise ValueError('Need a unique_attr to compute the client ID')
+
+        client_id = getattr(instance, self.unique_attr, None)
+        if not client_id:
```
Comment (nickpresta) [inline djsixpack/djsixpack.py:56]: If `client_id` can be `None` here, you want to use `is None`.

File: djsixpack/djsixpack.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            raise ValueError('Unique_attr {0} does not yield a usable identifier'.format(self.unique_attr))
+
+        return client_id
+
+    def participate(self, force=None, user_agent=None, ip_address=None):
+        if not self.host:
```
Comment (nickpresta) [inline djsixpack/djsixpack.py:62]: `is None`

File: djsixpack/djsixpack.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+            except TypeError:
+                raise ValueError('No alternatives defined')
+
+        session = self._get_session(user_agent, ip_address)
+        experiment_name = self._get_experiment_name()
+        resp = session.participate(experiment_name, self.alternatives, force)
+        return resp['alternative']['name']
+
+    def convert(self, kpi=None):
+        if not self.host:
```
Comment (nickpresta) [inline djsixpack/djsixpack.py:74]: here too

File: djsixpack/tests/class_tests.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+from django.test import TestCase
+
+from djsixpack.djsixpack import SixpackTest
+
+
+class ClientIdTest(TestCase):
+	
+	def test_client_id_property_calls_method(self):
+		tst = SixpackTest(Mock())
```
Comment (nickpresta) [inline djsixpack/tests/class_tests.py]: I think it is more readable to just call it `t` or `test`

File: djsixpack/tests/class_tests.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+		tst.get_client_id = Mock()
+		
+		tst.client_id
+		self.assertTrue(tst.get_client_id.called)
+		
+	def test_client_id_no_attr_set(self):
+		tst = SixpackTest(Mock())
+		tst.unique_attr = None
+		
+		try:
```
Comment (nickpresta) [inline djsixpack/tests/class_tests.py]:
```
with self.assertRaises(ValueError):
    tst.client_id
```

File: djsixpack/tests/class_tests.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+			tst.client_id
+		except ValueError:
+			pass
+		else:
+			self.fail('ValueError not raised')
+		
+	def test_client_id_no_usable_identifier(self):
+		tst = SixpackTest(Mock(pk=None))
+		
+		try:
```
Comment (nickpresta) [inline djsixpack/tests/class_tests.py]: same here

File: MANIFEST.in (hunk 8 above)
Comment (dlanger) [inline MANIFEST.in:3]: Added

File: djsixpack/__init__.py (hunk 9 above)
Comment (dlanger) [inline djsixpack/__init__.py]: True

File: setup.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
+from setuptools import setup, find_packages
+
+install_reqs = ["Django>=1.4", "sixpack-client==1.0.0"]
+
+setup(
+    name='django-sixpack',
+    version=__import__('djsixpack').__version__,
+    author="Dan Langer",
+    author_email='dan@waveapps.com',
```
Comment (nickpresta) [inline setup.py]: I thought this was supposed to be `opensource@` in the event you're no longer here:

https://github.com/waveaccounting/vero-python/blob/master/setup.py#L18

File: djsixpack/djsixpack.py (hunk 10 above)
Comment (dlanger) [inline djsixpack/djsixpack.py:56]: `client_id` being `None` (or anything `False`-y) is an error, so should trigger that exception.

File: djsixpack/djsixpack.py (hunk 11 above)
Comment (dlanger) [inline djsixpack/djsixpack.py:62]: I think I'm fine with `''` having the same behaviour as `None` - both are cases where the user is deliberately setting the value to something `None`-like, and should behave that way.

File: djsixpack/djsixpack.py (hunk 12 above)
Comment (dlanger) [inline djsixpack/djsixpack.py:74]: Same as above.

File: setup.py (hunk 16 above)
Comment (dlanger) [inline setup.py]: I had that originally, but interpreted @ashchristopher's comment upthread as that it should be my address as the main author. 

@ashchristopher @nduthoit - what's our standard here?

File: djsixpack/tests/class_tests.py (hunk 13 above)
Comment (dlanger) [inline djsixpack/tests/class_tests.py]: Agreed

File: setup.py (hunk 16 above)
Comment (ashchristopher) [inline setup.py]: Let me find out tomorrow during the TLG.

Comment (ashchristopher) [inline setup.py]: As a datapoint of one: https://github.com/dropbox/responses/blob/master/setup.py#L47

Comment (dlanger) [inline setup.py]: My confusion is over `author_email` - I think `author` should be the name of the primary author if there is one.

Comment (ashchristopher) [inline setup.py]: Ah - ok. I understand now - that seems reasonable.

Comment (nickpresta) [inline setup.py]: Yes. Author is all Dan, but email should be something more "general" than just one person, IMO.

File: setup.py (hunk 4 above)
Comment (dlanger) [inline setup.py]: Surprisingly, yes:

```
(sixpack)➜ dan-xubuntu ~/waveapps/django-sixpack [v0] python setup.py test
running test
running egg_info
creating django_sixpack.egg-info
writing requirements to django_sixpack.egg-info/requires.txt
writing django_sixpack.egg-info/PKG-INFO
writing top-level names to django_sixpack.egg-info/top_level.txt
writing dependency_links to django_sixpack.egg-info/dependency_links.txt
writing manifest file 'django_sixpack.egg-info/SOURCES.txt'
reading manifest file 'django_sixpack.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'django_sixpack.egg-info/SOURCES.txt'
running build_ext
Creating test database for alias 'default'...
...........
----------------------------------------------------------------------
Ran 11 tests in 0.017s

OK
Destroying test database for alias 'default'...
```

**Pull request merged**